### PR TITLE
FFT Added more options to start tracking

### PR DESCRIPTION
**Pull request by Jarelk**
**kwargs is love, **kwargs is life.

### CODE REVIEW

**Review by saberzero1 — DISMISSED** [review]

**Comment by Marieke-van-der-Plaat** [comment]
Volgens mij moesten Max & Rienk hier nog naar kijken

**Review by Marieke-van-der-Plaat — DISMISSED** [review]
Volgens mij moesten Rienk en/of Max hier nog specifiek naar kijken?

File: CameraProcessor/processor/pipeline/process_frames.py [hunk 1]
```diff
@@ -94,23 +94,41 @@ def process_message_queue(ws_client, framebuffer, re_identifier, re_id_data):
...
-            # Extract the features from this bounding box and store them in the data.
-            re_id_data.add_query_feature(track_elem.object_id, feature_map)
-
-            # Also store the map of the first box_id to the object_id.
-            re_id_data.add_query_box(track_elem.box_id, track_elem.object_id)
+            # If we have a cutout, extract features from it.
+            if track_elem.cutout is not None:
+                feature_map = re_identifier.extract_features_from_cutout(track_elem.cutout)
+            # If we don't have a cutout, but we have a frame id and box id, attempt to find the box in the buffer.
+            elif all(k is not None for k in [track_elem.frame_id, track_elem.box_id]):
```
Comment (RienkF) [inline CameraProcessor/processor/pipeline/process_frames.py:106]: Not completely sure about the logic here. The protocol makes it possible to send both a cutout and a frameid/boxid. Doing an if/else discards information which might be needed. In an ideal scenario I think the processor should try to get its own cutout based on frameid/boxid and compare the quality to this to that of the given cutout and create features based on the best quality.

File: CameraProcessor/processor/pipeline/process_frames.py [hunk 2]
```diff
@@ -94,23 +94,41 @@ def process_message_queue(ws_client, framebuffer, re_identifier, re_id_data):
...
+                    logging.error(error)
+                # Frame found in the buffer, extract features from the discovered bounding box.
+                else:
+                    feature_map = re_identifier.extract_features(stored_frame, stored_box)
+            else:
+                logging.error("Not enough information in the tracking message to start tracking object.")
+                feature_map = None
+
+            # If successfully extracted, send the feature_map to the orchestrator.
+            if feature_map is not None:
```
Comment (RienkF) [inline CameraProcessor/processor/pipeline/process_frames.py:123]: There were some talks about also sending a fail message if the feature map cant be created, so maybe include that in an else here?

File: CameraProcessor/processor/pipeline/reidentification/ireidentifier.py [hunk 3]
```diff
@@ -21,6 +21,14 @@ def extract_features(self, frame_obj, track_obj):
         """
         raise NotImplementedError("Extract features function not implemented")
 
+    def extract_features_from_cutout(self, cutout):
```
Comment (RienkF) [inline CameraProcessor/processor/pipeline/reidentification/ireidentifier.py:24]: I feel like a notImplementedError is not a great way to implement interfaces. I'd prefer formal interfaces:
https://realpython.com/python-interface/#formal-interfaces

File: CameraProcessor/processor/webhosting/start_command.py [hunk 4]
```diff
@@ -6,16 +6,30 @@
...
-            box_id (int): Id of the box that contains the object to be tracked.
-            object_id (int): Identifier to track the object with   .
-        """
-        self.frame_id = frame_id
-        self.box_id = box_id
-        self.object_id = object_id
+    It uses the dict __init__ to add the **kwargs or dict to this class. This class is, for all intents and purposes,
+    a dictionary with entries accessible like properties. Additional functionality can be added.
+
+    We expect some, but not all, of the following key-value pairs:
```
Comment (RienkF) [inline CameraProcessor/processor/webhosting/start_command.py:15]: It's probably way too late to start this conversation, but I thought we agreed on google docstring format. The styleguide specifies docstring should be descriptive-style, meaning phrases such as "we expect" are not allowed and should be "the function expects" or "the function requires", even more preferable is simply "expects parameters x and y"

Comment (saberzero1) [reply CameraProcessor/processor/webhosting/start_command.py:15]: I have a point for the stand up regarding writing standards. This is one of the cases I want to discuss.

File: CameraProcessor/processor/pipeline/process_frames.py (hunk 2 above)
Comment (Jarelk) [reply CameraProcessor/processor/pipeline/process_frames.py:123]: It is in there.

File: CameraProcessor/processor/pipeline/reidentification/ireidentifier.py (hunk 3 above)
Comment (Jarelk) [reply CameraProcessor/processor/pipeline/reidentification/ireidentifier.py:24]: It is currently how its implemented **everywhere** like this, so wrong place to bring this up

File: CameraProcessor/processor/pipeline/process_frames.py (hunk 1 above)
Comment (Jarelk) [reply CameraProcessor/processor/pipeline/process_frames.py:106]: That ideal scenario was not part of the scope of this issue. As it stands, the frameid/boxid is not needed if you already have a cutout. Comparing two feature maps is a whole different issue.

File: CameraProcessor/processor/pipeline/process_frames.py (hunk 2 above)
Comment (Jarelk) [reply CameraProcessor/processor/pipeline/process_frames.py:123]: I see what you mean. I will add an additional error message to send to orchestrator..

File: CameraProcessor/processor/pipeline/process_frames.py [hunk 5]
```diff
@@ -139,6 +157,17 @@ def send_boxes_to_orchestrator(ws_client, frame_obj, detected_boxes, tracked_box
     logging.info(client_message)
 
 
+def send_error_to_orchestrator(ws_client, error):
```
Comment (RienkF) [inline CameraProcessor/processor/pipeline/process_frames.py:160]: This should be in a i.e. client.py file

**Comment by Jarelk** [comment]
> Volgens mij moesten Rienk en/of Max hier nog specifiek naar kijken?

Nah

**Review by saberzero1 — APPROVED** [review]

**Review by JTeeuwissen — APPROVED** [review]

**Pull request merged**